docs/spec/connection-handshake/interfaces.md

```diff
@@ -0,0 +1,168 @@
+    expectedConsensusState = getConsensusState(consensusHeight)
+    expected = ConnectionEnd{INIT, desiredIdentifier, getCommitmentPrefix(), counterpartyClientIdentifier,
+                             clientIdentifier, counterpartyVersions}
+    version = pickVersion(counterpartyVersions)
+    connection = ConnectionEnd{state, counterpartyConnectionIdentifier, counterpartyPrefix,
+                               clientIdentifier, counterpartyClientIdentifier, version}
+    abortTransactionUnless(connection.verifyConnectionState(proofHeight, proofInit, counterpartyConnectionIdentifier, expected))
+    abortTransactionUnless(connection.verifyClientConsensusState(
+      proofHeight, proofConsensus, counterpartyClientIdentifier, consensusHeight, expectedConsensusState))
+    previous = provableStore.get(connectionPath(desiredIdentifier))
+    abortTransactionUnless(
+      (previous === null) ||
+      (previous.state === INIT &&
+        previous.counterpartyConnectionIdentifier === counterpartyConnectionIdentifier &&
+        previous.counterpartyPrefix === counterpartyPrefix &&
+        previous.clientIdentifier === clientIdentifier &&
+        previous.counterpartyClientIdentifier === counterpartyClientIdentifier &&
+        previous.version === version))
+    identifier = desiredIdentifier
+    state = TRYOPEN
+    provableStore.set(connectionPath(identifier), connection)
+    addConnectionToClient(clientIdentifier, identifier)
+}
+```
+
+
+### 2. Using abstractions
+
+We will use the following three abstractions to parametrize the `connOpenTry` handler:
+
+- The `Party` type:
+
+```typescript
+interface Party {
+  connectionIdentifier: Identifier
+  clientIdentifier: Identifier
+  prefix: CommitmentPrefix
+}
+```
+
+- A `Connection` type:
+
+```typescript
+interface Connection {
+  state: ConnectionState
+  localParty: Party
+  remoteParty: Party
+  version: string | []string
+}
+```
+
+- and a `Proof` type:
+
+```typescript
+interface Proof {
+  content: CommitmentProof,
+  height: uint64,
+}
+```
+
+##### New Handler (version 2)
+
+Now we redefine `connOpenTry` handler as follows:
+
+```typescript
+function connOpenTry(
+  e: Connection,
+  proofInit: Proof,
+  proofConsensus: Proof) {
+    abortTransactionUnless(validateConnectionIdentifier(e.remoteParty.connectionIdentifier))
+    abortTransactionUnless(proofConsensus.height <= getCurrentHeight())
+    expectedConsensusState = getConsensusState(proofConsensus.height)
+
+    // basic validation of connection parties
+    abortTransactionUnless(e.state === INIT)
+    abortTransactionUnless(e.remoteParty.prefix === getCommitmentPrefix())
+
+    abortTransactionUnless(e.remoteParty.verifyConnectionState(proofInit, e))
+    abortTransactionUnless(e.remoteParty.verifyClientConsensusState(
+      proofConsensus, proofConsensus.height, expectedConsensusState))
+    previous = provableStore.get(connectionPath(e.remoteParty.connectionIdentifier))
```

Comment:
I think that calling this previous is confusing. It is connectionEnd.

Reply:
We removed all `ConnectionEnd` types and now we only have `Party` and `Connection`. In the previous version of this document we were using `ConnectionEnd` and it looked like there was too much redundancy in data we were transmitting. See [this commit comment](https://github.com/informalsystems/ibc-rs/commit/1842ed964b41cd0a0e83e8c10e1c2ba4e4791a2d#r38238027) for more info.